### PR TITLE
add clusterview api to list managedcluster

### PR DESCRIPTION
**Pull request by zhiweiyin318**

### CODE REVIEW

File: pkg/proxyserver/cache/clusterview.go [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+		return
+	}
+	ac.allResources = resources
+	ac.synchronizeClusterRoleBindings(userSubjectRecordStore, groupSubjectRecordStore)
+	ac.lastState = currentState
+	klog.V(2).Infoln("synchronize...", ac.allResources, ac.allUsers, ac.allGroups)
+}
+
+// synchronizeRoleBindings synchronizes access over each role binding
+func (ac *AuthCache) synchronizeClusterRoleBindings(userSubjectRecordStore cache.Store, groupSubjectRecordStore cache.Store) {
```
Comment (qiujian16) [inline pkg/proxyserver/cache/clusterview.go:178]: More doc here please
1. how to handle role create/update
2. how to handle rolebinding create/update
3. how to handle role/rolebinding delete

**Review by qiujian16 — COMMENTED** [review]
I would expect ut for also for watch/ruleallow etc. We could reuse those in ocp apiserver

File: pkg/proxyserver/cache/clusterview.go (hunk 1 above)
Comment (zhiweiyin318) [reply pkg/proxyserver/cache/clusterview.go:184]: done

File: test/e2e/clusterview_test.go [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+
+		// prepare 3 clusters
+		cluster1, err := util.CreateManagedCluster(dynamicClient)
+		gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+		cluster2, err := util.CreateManagedCluster(dynamicClient)
+		gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+		cluster3, err := util.CreateManagedCluster(dynamicClient)
+		gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+
+		// Case 1: authorize cluster1, cluster2 to uer
```
Comment (qiujian16) [inline test/e2e/clusterview_test.go:150]: s/uer/user/g

File: test/e2e/clusterview_test.go [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+			return fmt.Errorf("validateClusterView: no resource %v", name)
+		}
+	}
+
+	return nil
+}
+
+var _ = ginkgo.Describe("Testing ClusterView", func() {
+	ginkgo.It("should get the managedClusters.clusterview successfully", func() {
+		// create clusterRole and clusterRoleBinding for user
```
Comment (qiujian16) [inline test/e2e/clusterview_test.go:135]: i think we should also have cases for creating/deleting clusterrole

File: pkg/proxyserver/rest/managedcluster/managedcluster.go [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+		return nil, errors.NewForbidden(clusterv1.Resource("managedclusters"), "", fmt.Errorf("unable to get mangedCluster without a user on the context"))
+	}
+	s.clusterCache.Get(name)
+	clusterList, err := s.lister.List(user, labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+	for _, cluster := range clusterList.Items {
+		if name == cluster.Name {
+			return s.client.ClusterV1().ManagedClusters().Get(ctx, name, *options)
```
Comment (qiujian16) [inline pkg/proxyserver/rest/managedcluster/managedcluster.go:122]: will this work? since the apigroup is different.

**Review by qiujian16 — COMMENTED** [review]
some nit, i think we also need watch e2e case as a followup

File: test/e2e/clusterview_test.go (hunk 2 above)
Comment (zhiweiyin318) [reply test/e2e/clusterview_test.go]: done

File: test/e2e/clusterview_test.go (hunk 3 above)
Comment (zhiweiyin318) [reply test/e2e/clusterview_test.go:154]: added

File: pkg/proxyserver/rest/managedcluster/managedcluster.go (hunk 4 above)
Comment (zhiweiyin318) [reply pkg/proxyserver/rest/managedcluster/managedcluster.go]: yes, it works.  I replace `s.client.ClusterV1().ManagedClusters().Get` with `s.clusterCache.Get()`.

File: test/e2e/clusterview_test.go (hunk 3 above)
Comment (zhiweiyin318) [reply test/e2e/clusterview_test.go:155]: watch case is added too

**Pull request merged**